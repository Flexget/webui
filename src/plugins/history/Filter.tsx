import React, { FC, useMemo } from 'react';
import { useInjectContent } from 'core/layout/AppBar/hooks';
import { Direction } from 'utils/query';
import { GroupByFields, SortByFields } from './types';
import FilterNav from './FilterNav';

interface Props {
  order: Direction;
  grouping: GroupByFields;
  sort: SortByFields;
  toggleOrder: () => void;
  handleChange: (key: string) => (e: any) => void;
}

const Filter: FC<Props> = props => {
  useInjectContent(useMemo(() => <FilterNav {...props} />, [props]));

  return null;
};

export default Filter;
