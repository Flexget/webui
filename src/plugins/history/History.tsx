import React, { FC, useRef, useEffect, useCallback } from 'react';
import { findDOMNode } from 'react-dom';
import { hot } from 'react-hot-loader/root';
import { useInjectPageTitle } from 'core/layout/AppBar/hooks';
import { NoPaddingWrapper } from 'common/styles';
import { useMergeState } from 'utils/hooks';
import { Direction } from 'utils/query';
import { Formik } from 'formik';
import InfiniteScroll from 'react-infinite-scroller';
import FilterNav from './FilterNav';
import HistoryList from './HistoryList';
import { GroupByFields, SortByFields, GetHistoryOptions } from './types';

const History: FC = () => {
  useInjectPageTitle('History');
  const scroll = useRef<InfiniteScroll>(null);
  const [options, setOptions] = useMergeState<GetHistoryOptions>({
    page: 1,
    grouping: GroupByFields.Time,
    sort: SortByFields.Time,
    order: Direction.Desc,
    task: '',
  });

  useEffect(() => {
    if (scroll.current && options.page === 1) {
      const node = findDOMNode(scroll.current); // eslint-disable-line react/no-find-dom-node

      if (node && 'scrollIntoView' in node) {
        node.scrollIntoView();
      }
      scroll.current.pageLoaded = 1;
    }
  }, [options.page]);

  const loadMore = useCallback((page: number) => setOptions({ page }), [setOptions]);

  return (
    <NoPaddingWrapper>
      <Formik
        initialValues={options}
        onSubmit={values =>
          setOptions({
            ...values,
            page: 1,
          })
        }
      >
        <FilterNav />
      </Formik>
      <HistoryList options={options} ref={scroll} loadMore={loadMore} />
    </NoPaddingWrapper>
  );
};

export default hot(History);
