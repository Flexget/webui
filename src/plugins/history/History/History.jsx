import React, { Component } from 'react';
import { findDOMNode } from 'react-dom';
import PropTypes from 'prop-types';
import HistoryList from 'plugins/history/HistoryList';
import FilterNav from 'plugins/history/FilterNav';
import { NoPaddingWrapper } from 'common/styles';

export default class HistoryPage extends Component {
  scroll = React.createRef();

  static propTypes = {
    getHistory: PropTypes.func.isRequired,
  };

  state = {
    grouping: 'time',
    sort: 'time',
    order: 'desc',
  };

  getHistory = page =>
    this.props.getHistory({
      page,
      sortBy: this.state.sort,
      order: this.state.order,
      task: this.state.task,
    });

  restartLoader = () => {
    findDOMNode(this.scroll.current).scrollIntoView(); // eslint-disable-line react/no-find-dom-node
    this.getHistory(1);
    this.scroll.current.pageLoaded = 1;
  };

  handleChange = key => event => this.setState({ [key]: event.target.value }, this.restartLoader);

  toggleOrder = () =>
    this.setState(
      ({ order }) => ({
        order: order === 'asc' ? 'desc' : 'asc',
      }),
      this.restartLoader,
    );

  render() {
    const { grouping } = this.state;

    return (
      <NoPaddingWrapper>
        <FilterNav
          handleChange={this.handleChange}
          toggleOrder={this.toggleOrder}
          {...this.state}
        />
        <HistoryList grouping={grouping} getHistory={this.getHistory} ref={this.scroll} />
      </NoPaddingWrapper>
    );
  }
}
