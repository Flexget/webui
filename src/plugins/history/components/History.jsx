import React, { Component } from 'react';
import { findDOMNode } from 'react-dom';
import PropTypes from 'prop-types';
import HistoryList from 'plugins/history/components/HistoryList';
import FilterNav from 'plugins/history/components/FilterNav';
import { NoPaddingWrapper } from 'components/styles';

export default class HistoryPage extends Component {
  static propTypes = {
    getHistory: PropTypes.func.isRequired,
  };

  scroll = React.createRef()

  list = React.createRef()

  state = {
    grouping: 'time',
    sort: 'time',
    order: 'desc',
  };

  getHistory = page => this.props.getHistory({
    page,
    sort_by: this.state.sort,
    order: this.state.order,
    task: this.state.task,
  });

  restartLoader = () => {
    findDOMNode(this.scroll).scrollIntoView(); // eslint-disable-line react/no-find-dom-node
    this.getHistory(1);
    this.scroll.pageLoaded = 1;
  }

  handleChange = key => event => this.setState({ [key]: event.target.value }, this.restartLoader)

  toggleOrder = () => this.setState(({ order }) => ({
    order: (order === 'asc' ? 'desc' : 'asc'),
  }), this.restartLoader)

  render() {
    const { grouping } = this.state;

    return (
      <NoPaddingWrapper>
        <FilterNav
          handleChange={this.handleChange}
          toggleOrder={this.toggleOrder}
          {...this.state}
        />
        <HistoryList
          grouping={grouping}
          getHistory={this.getHistory}
          ref={{
            listRef: this.list,
            scrollRef: this.scroll,
          }}
        />
      </NoPaddingWrapper>
    );
  }
}
