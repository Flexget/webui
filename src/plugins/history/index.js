import { createAsyncComponent } from 'utils/loading';
import registry from 'core/registry';
import saga from './state/saga';
import reducer from './state/reducer';

export default () =>
  registry.registerPlugin('history', {
    component: createAsyncComponent(() => import('plugins/history/History')),
    routeDisplayName: 'History',
    routeIcon: 'history',
    reducer,
    saga,
  });
