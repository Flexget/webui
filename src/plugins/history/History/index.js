import { connect } from 'react-redux';
import actions from '../state/actions';
import History from './History';

function mapDispatchToProps(dispatch) {
  return {
    getHistory: data => dispatch(actions.getHistory.request(data)),
  };
}

export default connect(null, mapDispatchToProps)(History);
export { History };
