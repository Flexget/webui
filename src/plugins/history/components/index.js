import { connect } from 'react-redux';
import { request } from 'utils/actions';
import { GET_HISTORY } from 'plugins/history/data/actions';
import History from './History';

function mapDispatchToProps(dispatch) {
  return {
    getHistory: data => dispatch(request(GET_HISTORY, data)),
  };
}

export default connect(null, mapDispatchToProps)(History);
export { History };
