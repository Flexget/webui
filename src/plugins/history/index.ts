import { createAsyncComponent } from 'utils/loading';
import { registerPlugin } from 'core/plugins/registry';
import HistoryIcon from '@material-ui/icons/History';

export default () =>
  registerPlugin('/history', {
    component: createAsyncComponent(() =>
      import(
        /* webpackChunkName: 'HistoryPlugin' */
        'plugins/history/History'
      ),
    ),
    routeDisplayName: 'History',
    routeIcon: HistoryIcon,
  });
