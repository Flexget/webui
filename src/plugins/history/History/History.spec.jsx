import React from 'react';
import { shallow } from 'enzyme';
import { History } from 'plugins/history/History';

describe('plugins/history/components', () => {
  it('renders correctly', () => {
    const tree = shallow(<History getHistory={jest.fn()} />);
    expect(tree).toMatchSnapshot();
  });
});
