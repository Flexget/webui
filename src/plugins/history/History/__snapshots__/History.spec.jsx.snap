// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/history/components renders correctly 1`] = `
<Styled(div)>
  <FilterNav
    grouping="time"
    handleChange={[Function]}
    order="desc"
    sort="time"
    toggleOrder={[Function]}
  />
  <ForwardRef
    getHistory={[Function]}
    grouping="time"
  />
</Styled(div)>
`;
